src/shogun/machine/GLM.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Tej Sukhatme
+ */
+
+
+#include <shogun/lib/config.h>
+
+#include <shogun/lib/common.h>
+#include <shogun/machine/Machine.h>
+#include <shogun/machine/LinearMachine.h>
+#include <shogun/machine/IterativeMachine.h>
+#include <shogun/mathematics/RandomMixin.h>
+#include <shogun/distributions/Distribution.h>
+
+namespace shogun
+{
+    enum GLM_DISTRIBUTION
+	{
+        //TODO GAUSSIAN,
+        //TODO BINOMIAL,
+        //TODO GAMMA,
+        //TODO SOFTPLUS,
+        //TODO PROBIT,
+        POISSON
+	};
+
+class DotFeatures;
+class Features;
+class RegressionLabels;
+
+/** @brief Class for estimating regularized generalized linear models (GLM).
+ *   The regularized GLM minimizes the penalized negative log likelihood:
+ *      .. math::
+ *
+ *		\\min_{\\beta_0, \\beta} \\frac{1}{N}
+ *		\\sum_{i = 1}^N \\mathcal{L} (y_i, \\beta_0 + \\beta^T x_i)
+ *		+ \\lambda [ \\frac{1}{2}(1 - \\alpha) \\mathcal{P}_2 +
+ *						\\alpha \\mathcal{P}_1 ]
+ *
+ *	where :math:`\\mathcal{P}_2` and :math:`\\mathcal{P}_1` are the generalized
+ *	L2 (Tikhonov) and generalized L1 (Group Lasso) penalties, given by:
+ *
+ *	.. math::
+ *
+ *		\\mathcal{P}_2 = \\|\\Gamma \\beta \\|_2^2 \\
+ *		\\mathcal{P}_1 = \\sum_g \\|\\beta_{j,g}\\|_2
+ *
+ *	where :math:`\\Gamma` is the Tikhonov matrix: a square factorization
+ *	of the inverse covariance matrix and :math:`\\beta_{j,g}` is the
+ *	:math:`j` th coefficient of group :math:`g`.
+ *
+ *	The generalized L2 penalty defaults to the ridge penalty when
+ *	:math:`\\Gamma` is identity.
+ *
+ *	The generalized L1 penalty defaults to the lasso penalty when each
+ *	:math:`\\beta` belongs to its own group.
+ *
+ * */
+class GLM : public IterativeMachine<LinearMachine>, public RandomMixin<Distribution>
```

Comment:
the problem with multi-inheritance is that this is not exposeable to JVM.... as multi-inheritance is not supported :(

Reply:
> the problem with multi-inheritance is that this is not exposeable to JVM.... as multi-inheritance is not supported :(

Oh okay, so, what's the alternative? I need the `IterativeMachine` to run the several Iterations and I need the `RandomMixin` to initialize the parameters with random values accoring to a normal distribution in the `init_model()` function

Reply:
But why `RandomMixin<Distribution>`?

Reply:
I always thought that if a class is not exposed to swig, then there are no problems with multi inheritance?

Reply:
but i guess for the time being this will be exposed.
still i'm not so sure if i get why the Distribution interface required here...

Reply:
note i believe that this should be `RandomMixin<IterativeMachine<LinearMachine>>` since i really dont see how GLM is a distribution itself...

Reply:
I used the `Distribution` interface as I had seen it get used elsewhere.

Yes, I had tried that before but It was throwing a few errors I wasn't able to figure out then. Since multi inheritance is a problem, I will be implementing it using `RandomMixin<IterativeMachine<LinearMachine>>`

Reply:
++ on removing distribution. IMO the multiple inheritance is not a problem, but that should be resolved anyways with your change

Reply:
Yes, I have removed it.